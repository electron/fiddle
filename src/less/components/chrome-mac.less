.darwin {
  .chrome {
    text-align: center;
    background-color: @background-1;
    border-bottom: @border;
    -webkit-app-region: drag;
    height: 24px;
    font-size: 1.4rem;
    line-height: 15px;

    small,
    p,
    span {
      user-select: none;
    }
  }

  .big_sur_or_later {
    height: 30px !important;
  }
}
