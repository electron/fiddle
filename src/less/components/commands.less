@import "../variables.less";

header {
  box-shadow: 0 0 0 1px rgba(16, 22, 26, 0.2), 0 0 0 rgba(16, 22, 26, 0), 0 1px 1px rgba(16, 22, 26, 0.4);
  background-color: @background-3;
  -webkit-app-region: drag;
}

.disabled-menu-tooltip {
  display: flex;
  width: 100%;

  .bp3-popover-target {
    width: 100%;
  }
}

.commands.is-mac {
  padding:
    10px
    calc(100vw - (env(titlebar-area-x, 0) + env(titlebar-area-width, 0)))
    10px
    env(titlebar-area-x, 0);
}

.commands {
  display: flex;
  justify-content: space-between;
  height: 50px;
  padding:
    10px;
  box-sizing: border-box;
  font-family: @fonts-common;

  button,
  input,
  select {
    -webkit-app-region: no-drag;
    margin-bottom: 0;
    white-space: nowrap;
  }

  select {
    width: 100px;
    margin-left: 5px;
  }

  fieldset {
    border: 0;
    padding: 0;
    margin: 0;
    min-width: 0;
  }

  & > div {
    display: flex;

    &:last-of-type {
      margin-right: 10px;

      .bp3-button-group {
        margin-left: 5px;
      }
    }
  }

  .bp3-control-group {
    margin-right: 0.5rem;
  }

  .address-bar {
    width: 420px;
    transition: width 0.4s;

    input,
    span {
      -webkit-app-region: no-drag;
    }

    &.empty {
      width: 300px;
    }

    &:focus-within {
      span,
      input {
        border-color: @foreground-2;
      }
    }
  }

  & > .title {
    align-self: stretch;
    flex-shrink: 0;

    flex-flow: column nowrap;
    justify-content: center;
    user-select: none;

    margin: 0 10px;
  }
}
