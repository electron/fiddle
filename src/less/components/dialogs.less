@import "../variables.less";

.dialogs {
  z-index: 10;
  position: fixed;
  height: 100vh;
  width: 100vw;
  pointer-events: none;

  * {
    pointer-events: all;
  }
}

.dialog {
  background-color: @background-3;
  border: @border;
  color: @text-color-1;
  border-radius: 6px;
  padding: 10px;
  z-index: 100;
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  animation: slide-up 0.4s ease;

  p {
    margin-bottom: 1.25rem;
    font-size: 1.25rem;
  }

  code {
    font-size: 12px;
    background: rgba(255, 255, 255, 0.1);
    padding: 2px;
    line-height: 12px;
  }

  &-buttons {
    margin-top: 1.25rem;
  }

  &.centered {
    position: absolute;
    width: 300px;
    left: calc(50% - (300px / 2));
    top: 100px;
  }

  &.arrow::after,
  &.arrow::before {
    border: solid transparent;
    content: " ";
    height: 0;
    width: 0;
    position: absolute;
    pointer-events: none;
  }

  &.arrow::after {
    border-color: rgba(136, 183, 213, 0);
    border-width: 9px;
  }

  &.arrow::before {
    border-color: rgba(194, 225, 245, 0);
    border-width: 10px;
  }

  &.arrow-top::after,
  &.arrow-top::before {
    bottom: 100%;
    left: 50%;
  }

  &.arrow-top::after {
    border-bottom-color: @background-3;
    margin-left: -9px;
  }

  &.arrow-top::before {
    border-bottom-color: @border-color-1;
    margin-left: -10px;
  }

  &.arrow-bottom::after,
  &.arrow-bottom::before {
    top: 100%;
    left: 50%;
  }

  &.arrow-bottom::after {
    border-top-color: @background-3;
    margin-left: -9px;
  }

  &.arrow-bottom::before {
    border-top-color: @border-color-1;
    margin-left: -10px;
  }

  &.arrow-left::after,
  &.arrow-left::before {
    right: 100%;
    top: 20px;
  }

  &.arrow-left::after {
    border-right-color: @background-3;
    margin-top: -9px;
  }

  &.arrow-left::before {
    border-right-color: @border-color-1;
    margin-top: -10px;
  }

  &.arrow-right::after,
  &.arrow-right::before {
    left: 100%;
    top: 20px;
  }

  &.arrow-right::after {
    border-left-color: @background-3;
    margin-top: -9px;
  }

  &.arrow-right::before {
    border-left-color: @border-color-1;
    margin-top: -10px;
  }

  button {
    color: #1e2527;
    float: right;
    margin-left: 9px;
    margin-bottom: 0;
  }

  a {
    cursor: pointer;
  }

  h4 {
    margin-bottom: 0;
  }

  p:last-of-type {
    margin-bottom: 0;
  }

  input {
    color: rgba(159, 234, 249, 0.5);
    height: 28px;
    margin-top: 10px;
    background-color: @background-1;
    border: 1px solid @border-color-2;
    outline: 0 solid transparent;

    &:focus {
      border-color: @foreground-2;
    }
  }
}

.dialogDrop {
  position: fixed;
  z-index: 100;
  top: 0;
  left: 0;
  width: 100vw;
  height: 100vh;
  background: black;
  transition: opacity 0.4s ease-in-out;
  pointer-events: all;
  opacity: 0.5;
}

.dialog-add-version {
  .bp3-file-input {
    width: 100%;
  }

  .bp3-callout {
    margin-top: 1rem;
  }
}

select {
  font-size: 12px;
  height: 28px;
  background-color: white;
  line-height: 15px;
}

@keyframes slide-up {
  0% {
    opacity: 0;
    transform: translateY(20px);
  }

  100% {
    opacity: 1;
    transform: translateY(0);
  }
}
