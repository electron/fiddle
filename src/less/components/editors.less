.mosaic-tile {
  animation: fadein 0.15s;
}

.mosaic-window-toolbar  {
  & > div {
    display: flex;
    justify-content: space-between;
    width: 100%;
    padding: 0 5px 0 5px;

    .mosaic-controls {
      transform: scale(0.75);
      transform-origin: right;
    }

    h5 {
      margin: 4px 0 0 0;
    }

    button {
      margin-left: 8px;
    }
  }
}

// TODO: support new file
// update if list of Editor ID changes
@editor-ids: main\.js, renderer\.js, index\.html, preload\.js, styles\.css;

// takes each editor ID and increase its z-index if the parent Mosaic root
// has focused__id class for that specific id.
each(@editor-ids, {
  .focused__@{value}.mosaic .mosaic-window.@{value} {
    z-index: 2;
  }
});

.mosaic-window {
  z-index: 1;
}

@keyframes fadein {
  from { opacity: 0; }
  to { opacity: 1; }
}
