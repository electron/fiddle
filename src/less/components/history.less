.revision-list {
  max-height: 400px;
  overflow-y: auto;
}

.revision-list ul {
  list-style-type: none;
  padding: 0;
  margin: 0;
}

.revision-item {
  padding: 10px;
  margin-bottom: 5px;
  border-radius: 3px;
  cursor: pointer;
  transition: background-color 0.2s ease;
  border-left: 3px solid #106ba3;
}

.revision-item:hover {
  background-color: rgba(167, 182, 194, 0.3);
}

.revision-content {
  display: flex;
  flex-direction: column;
}

.revision-icon {
  margin-right: 8px;
}

.sha-label {
  font-size: 12px;
  color: #738694;
  margin-left: 10px;
  font-family: monospace;
}

.revision-details {
  display: flex;
  justify-content: space-between;
  align-items: center;
  margin-top: 5px;
}

.revision-date {
  color: #738694;
  font-size: 12px;
}

.revision-changes {
  display: flex;
  gap: 5px;
}

.history-loading {
  display: flex;
  flex-direction: column;
  align-items: center;
  padding: 20px;
}
