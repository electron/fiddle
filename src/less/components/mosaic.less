@import '../variables.less';

.fiddle {
  .mosaic-window-toolbar {
    background: var(--background-1);
    box-shadow: inset 0 0 100px 100px rgba(53, 53, 53, 0.1);
    border-top: 1px solid rgba(255, 255, 255, 0.1);
  }

  .mosaic-window-toolbar.draggable:hover {
    box-shadow: inset 0 0 100px 100px rgba(255, 255, 255, 0.1);
  }

  .mosaic-window-title {
    color: var(--text-color-3);
  }

  button.mosaic-default-control {
    padding: 0;
    height: 20px;
    width: 20px;
    margin: 5px;
    background-color: #314246;
    border: 0.1rem solid #222b35;
  }

  button.mosaic-default-control:hover {
    background-color: #131d1f;
  }

  .mosaic-window-body {
    display: flex;
    flex-direction: row;
    flex-wrap: nowrap;
    align-items: stretch;
    align-content: stretch;
    overflow: visible;
    z-index: 4;
    background-color: @background-4;
  }

  .mosaic.mosaic-blueprint-theme {
    background: #bdbdbd15;
  }

  .editorContainer,
  .monaco-editor {
    width: 100%;
    height: 100%;
  }

  .editorContainer {
    background: @background-1;
  }
}
