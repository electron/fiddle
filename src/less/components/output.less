@import '../variables.less';

.output {
  font-size: 12px;
  font-family: @fonts-common;
  font-weight: 600;
  top: 10vh;
  padding: 0;
  height: 100%;
  -webkit-app-region: no-drag;

  .monaco-editor {
    color: var(--text-color-1);
    box-shadow: inset 0 0 100px 100px rgba(0, 0, 0, 0.15);
    border-top: @border;

    .margin,
    .monaco-editor-background {
      background: var(--background-2);
    }
  }
}
