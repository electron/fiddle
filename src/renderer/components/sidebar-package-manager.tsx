import * as React from 'react';

import { Button, MenuItem, Tree, TreeNodeInfo } from '@blueprintjs/core';
import { Suggest } from '@blueprintjs/select';
import { autorun } from 'mobx';
import { observer } from 'mobx-react';
import pDebounce from 'p-debounce';
import semver from 'semver';

import { npmSearch } from '../npm-search';
import { AppState } from '../state';
interface IState {
  suggestions: Array<AlgoliaHit>;
  versionsCache: Map<string, string[]>;
}

interface IProps {
  appState: AppState;
}

// See full schema: https://github.com/algolia/npm-search#schema
interface AlgoliaHit {
  name: string;
  version: string;
  versions: Record<string, string>;
  _highlightResult: {
    name: {
      value: string;
    };
  };
}

@observer
export class SidebarPackageManager extends React.Component<IProps, IState> {
  constructor(props: IProps) {
    super(props);
    this.state = {
      suggestions: [],
      versionsCache: new Map(),
    };
  }

  public componentDidMount() {
    autorun(async () => {
      await this.refreshVersionsCache();
      this.coerceInvalidVersionNumbers();
    });
  }

  public addModuleToFiddle = (item: AlgoliaHit) => {
    const { appState } = this.props;
    appState.modules.set(item.name, item.version);
    // copy state so react can re-render
    this.state.versionsCache.set(item.name, Object.keys(item.versions));
    const newCache = new Map(this.state.versionsCache);
    this.setState({ suggestions: [], versionsCache: newCache });
  };

  public render() {
    return (
      <div className="package-tree">
        <h5>Modules</h5>
        <Suggest
          fill={true}
          inputValueRenderer={() => ''}
          items={this.state.suggestions}
          itemRenderer={(item, { modifiers, handleClick }) => (
            <MenuItem
              active={modifiers.active}
              key={item.name}
              text={
                <span
                  className="package-manager-result"
                  dangerouslySetInnerHTML={{
                    __html: item._highlightResult.name.value,
                  }}
                />
              }
              onClick={handleClick}
            />
          )}
          noResults={<em>Search for modules here...</em>}
          onItemSelect={this.addModuleToFiddle}
          onQueryChange={pDebounce(async (query) => {
            if (query !== '') {
              const { hits } = await npmSearch.search(query);
              this.setState({
                suggestions: hits,
              });
            } else {
              this.setState({
                suggestions: [],
              });
            }
          }, 200)}
          popoverProps={{ minimal: true, usePortal: false, fill: true }}
          resetOnClose={false}
          resetOnSelect={true}
        />
        <Tree contents={this.getModuleNodes()} />
      </div>
    );
  }

  /**
   * Takes the module map and returns an object
   * conforming to the BlueprintJS tree schema.
   * @returns TreeNodeInfo[]
   */
  private getModuleNodes = (): TreeNodeInfo[] => {
    const values: TreeNodeInfo[] = [];
    const { appState } = this.props;
    for (const [pkg, activeVersion] of appState.modules.entries()) {
      values.push({
        id: pkg,
        label: pkg,
        secondaryLabel: (
          <div>
            <select
              className="package-tree-version-select"
              name={pkg}
              value={activeVersion}
              onChange={({ target }) =>
                appState.modules.set(target.name, target.value)
              }
            >
              {this.state.versionsCache.get(pkg)?.map((version) => (
                <option key={version}>{version}</option>
              ))}
            </select>

            <Button
              minimal
              icon="remove"
              onClick={() => appState.modules.delete(pkg)}
            />
          </div>
        ),
      });
    }

    return values;
  };

  /**
   * Attempt to fetch the list of all versions for
   * all installed modules. We need this list of versions
   * for the version selector.
   */
  private refreshVersionsCache = async () => {
    const { modules } = this.props.appState;

    for (const pkg of modules.keys()) {
      if (!this.state.versionsCache.has(pkg)) {
        const { hits } = await npmSearch.search(pkg);
        const firstMatch = hits[0];
        if (firstMatch === undefined || !firstMatch.versions) {
          console.warn(
            `Attempted to fetch version list for ${pkg} from Algolia but failed!`,
          );
        } else {
          this.state.versionsCache.set(
            firstMatch.name,
            Object.keys(firstMatch.versions),
          );
          // React won't re-render when we're adding
          // elements to an ES6 Map element. Trigger
          // the re-render manually.
          this.forceUpdate();
        }
      }
    }
  };

  /**
   * Coerces any invalid semver versions to the latest
   * version detected in the versionsCache. This is particularly
   * useful for loading gists created with older versions of
   * Fiddle that have wildcard (*) versions in their package.json
   *
   * This function should only be run after the versions cache
   * is updated so we have an updated list of all deps and their
   * versions.
   */
  private coerceInvalidVersionNumbers = () => {
    const { modules } = this.props.appState;

    for (const [pkg, version] of modules.entries()) {
      if (!semver.valid(version)) {
        const packageVersions = this.state.versionsCache.get(pkg);

        if (Array.isArray(packageVersions) && packageVersions.length > 0) {
          const latestVersion = packageVersions[packageVersions.length - 1];
          modules.set(pkg, latestVersion);
        } else {
          console.warn(
            `Attempted to coerce latest version for package '${pkg}' but failed.`,
          );
        }
      }
    }
  };
}
