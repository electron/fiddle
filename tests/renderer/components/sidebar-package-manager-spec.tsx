import * as React from 'react';

import { Button } from '@blueprintjs/core';
import { mount, shallow } from 'enzyme';

import { SidebarPackageManager } from '../../../src/renderer/components/sidebar-package-manager';

jest.mock('../../../src/renderer/npm-search', () => ({
  npmSearch: {
    // this is just enough mocking to hit the right code paths
    // by stubbing out the npmSearch utility.
    search: () => ({
      hits: [
        {
          versions: ['1.0.0'],
        },
      ],
    }),
  },
}));

describe('SidebarPackageManager component', () => {
  let store: any;
  beforeEach(() => {
    store = {
      modules: new Map<string, string>([['cow', '1.0.0']]),
    };
  });

  it('renders', () => {
    const wrapper = shallow(<SidebarPackageManager appState={store} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('can add a package', () => {
    const wrapper = shallow(<SidebarPackageManager appState={store} />);
    const instance = wrapper.instance();
    instance.state = {
      suggestions: [],
      versionsCache: new Map(),
    };

    (instance as any).addModuleToFiddle({
      name: 'say',
      version: '2.0.0',
      versions: ['1.0.0', '2.0.0'],
    });

    expect(
      Array.from((store.modules as Map<string, string>).entries()),
    ).toMatchSnapshot([
      ['cow', '1.0.0'],
      ['say', '2.0.0'],
    ]);
  });

  it('can remove a package', async () => {
    const wrapper = mount(<SidebarPackageManager appState={store} />);

    const instance = wrapper.instance();
    instance.state = {
      suggestions: [],
      versionsCache: new Map(),
    };

    const btn = wrapper.find(Button);
    btn.simulate('click');
    expect((store.modules as Map<string, string>).size).toBe(0);

    // dumb timeout for setState to finish running
    await new Promise((resolve) => setTimeout(resolve, 1000));
  });
});
