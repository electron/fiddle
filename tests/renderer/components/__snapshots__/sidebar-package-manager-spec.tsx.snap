// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SidebarPackageManager component can add a package 1`] = `
Array [
  Array [
    "cow",
    "1.0.0",
  ],
  Array [
    "say",
    "2.0.0",
  ],
]
`;

exports[`SidebarPackageManager component renders 1`] = `
<div
  className="package-tree"
>
  <h5>
    Modules
  </h5>
  <Blueprint4.Suggest2
    closeOnSelect={true}
    fill={true}
    inputValueRenderer={[Function]}
    itemRenderer={[Function]}
    items={Array []}
    noResults={
      <em>
        Search for modules here...
      </em>
    }
    onItemSelect={[Function]}
    onQueryChange={[Function]}
    openOnKeyDown={false}
    popoverProps={
      Object {
        "minimal": true,
        "usePortal": false,
      }
    }
    resetOnClose={false}
    resetOnSelect={true}
  />
  <Blueprint4.Tree
    contents={
      Array [
        Object {
          "id": "cow",
          "label": "cow",
          "secondaryLabel": <div>
            <select
              className="package-tree-version-select"
              name="cow"
              onChange={[Function]}
              value="1.0.0"
            />
            <Blueprint4.Button
              icon="remove"
              minimal={true}
              onClick={[Function]}
            />
          </div>,
        },
      ]
    }
  />
</div>
`;
